File: packages/kotlin-tex-core/src/jvmBenchmark/kotlin/com/nouprax/tex/core/benchmark/Benchmark.kt
package com.nouprax.tex.core.benchmark

import com.nouprax.tex.core.Document
import com.nouprax.tex.core.model.CompileMode
import com.nouprax.tex.core.model.CompileOptions
import kotlin.time.measureTime

private const val WARMUP_COUNT = 3
private const val REPEAT_COUNT = 10

private fun benchmark(
    workload: String,
    source: String,
    mode: CompileMode,
) {
    val options = CompileOptions(mode)
    repeat(WARMUP_COUNT) { Document.compile(source, options) }

    val durations =
        (0 until REPEAT_COUNT)
            .map { measureTime { Document.compile(source, options) } }
            .sorted()
    val medianNanoseconds = durations[REPEAT_COUNT / 2].inWholeNanoseconds
    val peakRssKib =
        java.lang.management.ManagementFactory
            .getMemoryMXBean()
            .heapMemoryUsage.used / 1024
    println(
        "benchmark runtime=kotlin-jvm boundary=native_compile_and_value_copy workload=$workload " +
            "bytes=${source.encodeToByteArray().size} warmup=$WARMUP_COUNT repeats=$REPEAT_COUNT " +
            "median_ns=$medianNanoseconds heap_used_kib=$peakRssKib",
    )
}

public fun main() {
    val paragraph = "The 42 rows total 3.14 units, and the galley keeps flowing.\n"
    benchmark("large_document", paragraph.repeat(2_000), CompileMode.DOCUMENT)
    benchmark("math_spacing", "a\\,b\\:c\\;d\\!e\\quad f\\qquad g\\ h".repeat(200), CompileMode.MATH_DISPLAY)
}

File: packages/es-tex-core/scripts/benchmark.mjs
import { Document } from "../dist/index.js";

const warmupCount = 3;
const repeatCount = 10;

function benchmark(workload, source, mode) {
    const options = { mode };
    for (let index = 0; index < warmupCount; index += 1) Document.compile(source, options);

    const durations = [];
    for (let index = 0; index < repeatCount; index += 1) {
        const start = process.hrtime.bigint();
        Document.compile(source, options);
        durations.push(Number(process.hrtime.bigint() - start));
    }
    durations.sort((left, right) => left - right);
    const medianNanoseconds = durations[Math.trunc(repeatCount / 2)];
    const heapUsedKib = Math.trunc(process.memoryUsage().heapUsed / 1024);
    console.log(
        `benchmark runtime=es-node boundary=wasm_compile_and_value_copy workload=${workload} ` +
            `workload_version=1 ` +
            `bytes=${Buffer.byteLength(source, "utf8")} warmup=${warmupCount} repeats=${repeatCount} ` +
            `median_ns=${medianNanoseconds} heap_used_kib=${heapUsedKib}`
    );
}

const paragraph = "The 42 rows total 3.14 units, and the galley keeps flowing.\n";
benchmark("large_document", paragraph.repeat(2000), "document");
benchmark("math_spacing", "a\\,b\\:c\\;d\\!e\\quad f\\qquad g\\ h".repeat(200), "mathDisplay");
